1 Debugging
As you work through this problem set, whenever you encounter an error, please include a comment explaining what the error was and how you later resolved it. You can add these explanations at any relevant place in the file. Example:
>>> print("Hello, World!")
>>> """
>>> I encountered the following eror: "SyntaxError: unterminated
>>> string literal (detected at line 1) when I wrote
>>> print("Hello, World!)". So I added the missing " at the end
>>> and the code printed it successfully.

for problem set 2.3, it kept telling me that sub_task2 was undefined when I tried to slice it and then I realized that I never set it equal to something. I fixed it by setting it equal to something and then slicing it.
For most of these sections, instead of getting an error, it just wouldn't print anything. So not a lot of errors, just a lot of nothing.
got this error on the last problem TypeError: not all arguments converted during string formatting

2.1 Making a List Variable
Create a variable (name it anything you want, but make it descriptive!) that is assigned to a list with numbers 0 through 20. You should not write each number manually.
>>> whateverNameYouWant = # Your code here
>>> <br> print(whateverNameYouWant)
<br>[0, 1, 2, ..., 20] # It should print all numbers 1-20 in a list

In [204]:
task_1 = []
for nums in range(0,21):
    task_1.append(nums)
    
print(task_1)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]



Write a function that will take in your list from Part 2.1 and square each element in the list. Assign the result to a new variable.
>>> whateverNameYouWant = # Your code from Part 2.1
>>><br>def squareList():
>>> <br># Your code here
<br>
>>> <br>anotherNameYouWant = squareList(list)
>>><br> print(anotherNameYouWant)
<br>[0, 1, 4, ..., 400]

In [205]:
task_2 = []
for nums_2 in task_1:
    squares = nums_2 ** 2
    task_2.append(squares)

print(task_2)


[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


Write a function that takes in your new list from Part 2.2 and returns the first 15 elements of the list using slicing.
>>> anotherNameYouWant = squareList(list)
>>> <br>first_fifteen_elements(anotherNameYouWant)
<br>[0, 1, 4, ..., 196]

In [206]:

sub_task2= task_2[0:15]
print(sub_task2)


[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196]


2.4 Striding
Write a function that takes in your list from Part 2.2 and returns every 5th element from the list using striding.
>>> <br>anotherNameYouWant = squareList(list)
>>> <br>every_fifth_element(anotherNameYouWant)
<br>[16, 81, 196, 361]

In [207]:


print(task_2[4::5])

#stride_task2= task_2[4::5]
#print(stride_task2)
#I thought I needed to do what I did above but learned that I just needed to print it. 

[16, 81, 196, 361]


2.5 Slicing & Striding
Write a function that takes your list from Part 2.2, slices the last 30 elements of the list, and then returns every 3rd element from that list in reverse order.
>>> <br>anotherNameYouWant = squareList(list)
>>> <br>fancy_function(anotherNameYouWant)
<br>[400, 289, 196, 121, 64, 25, 4]


In [208]:
#don't know why it is saying 30 but I will try to get the same numbers as the example did.
task_2 = []
for nums_2 in task_1:
    squares = nums_2 ** 2
    task_2.append(squares)

print(task_2)

task_2.reverse()

print(task_2[::3])

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]
[400, 289, 196, 121, 64, 25, 4]


3 3D Lists
<br>3.1 Creating a 5x5 2D List
<br>Write a function that uses nested for loops to create a 5x5 2D list where the numbers 1 through 25 are stored sequentially. Assign the result to a new variable.
>>> <br>matrix = create_2d_list()
>>> <br>print(matrix)
<br>[[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15],
<br>[16, 17, 18, 19, 20], [21, 22, 23, 24, 25]]


In [209]:
nestedList = [list1, list2] #list3, list4, list5]
list1 = []
for number1 in range(1,6):
    list1.append(number1)
    
    list2 = []
    for number2 in range(6,11):
        list2.append(number2)
    list = []
  
print(nestedList)
#this is just me trying to understand how to do it without nested for loops

[['?', 22, 23, '?', 25], [6, 7, 8, 9, 10]]


In [210]:
nestedList = []
for num in range(0,5):
    list1 = []
    for num2 in range(0,5):
        function = num * 5 + num2 + 1 #figured out I had to add a 1 so it wouldn't start at zero, I originally had 0,6 then changed it to 0,5 
        list1.append(function)
    nestedList.append(list1)

print(nestedList)

[[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15], [16, 17, 18, 19, 20], [21, 22, 23, 24, 25]]


3.2 Replacing 2D List with Multiples of 3
<br>With the 2D list you created in Part 3.1, write a function that will replace all multiples of 3 with the character “?”.
>>><br> matrix = create_2d_list()
>>> <br>modified_2d_list(matrix)
<br>[[1, 2, ‘?’, 4, 5],
<br>[‘?’, 7, 8, ‘?’, 10],
<br>[11, ‘?’, 13, 14, ‘?’],
<br>[16, 17, ‘?’, 19, 20],
<br>[‘?’, 22, 23, ‘?’, 25]]


In [211]:
nestedList = []
for num in range(0,5):
    list1 = []
    for num2 in range(0,5):
        function = num * 5 + num2 + 1 
        list1.append(function)
    nestedList.append(list1)


print(nestedList)

[[1, 2, 3, 4, 5], [6, 7, 8, 9, 10], [11, 12, 13, 14, 15], [16, 17, 18, 19, 20], [21, 22, 23, 24, 25]]


In [212]:
nestedList = []
for num in range(0,5):
    list1 = []
    for num2 in range(0,5):
        function = num * 5 + num2 + 1 
        list1.append(function)
    nestedList.append(list1)

def nestedMatrix(): 
    for list in range(len(nestedList)):
        for numbers in range(len(nestedList[list])):
            if nestedList[list][numbers] % 3 == 0:
                nestedList[list][numbers] = "?"
    return nestedList


nestedMatrix()


[[1, 2, '?', 4, 5],
 ['?', 7, 8, '?', 10],
 [11, '?', 13, 14, '?'],
 [16, 17, '?', 19, 20],
 ['?', 22, 23, '?', 25]]

3.3 Summing None-’ ?’ Elements
<br>Assign the result of your function from Part 3.2 to a variable. Write a function that will iterate through the new 2D list variable and return the sum of all the elements that are not “?”. Hint: Try using “!=”.
>>><br>matrix = create_2d_list()
>>> <br>new_matrix = modified_2d_list(matrix)
>>> <br>sum_non_question_elements(new_matrix)
<br>217

In [213]:
nestedList = []
for num in range(0,5):
    list1 = []
    for num2 in range(0,5):
        function = num * 5 + num2 + 1 
        list1.append(function)
    nestedList.append(list1)

def nestedMatrix(): 
    for list in range(len(nestedList)):
        for numbers in range(len(nestedList[list])):
            if nestedList[list][numbers] % 3 == 0:
                nestedList[list][numbers] = "?"
    return nestedList

def SummingFunction(listInput):
    sum = 0
    for list in range(len(listInput)):
        for numbers in range(len(listInput[list])):
            print(numbers)
            if numbers != "?":
                sum += numbers
    return sum

SummingFunction(nestedList)
#ran into this issue


0
1
2
3
4
0
1
2
3
4
0
1
2
3
4
0
1
2
3
4
0
1
2
3
4


50

In [214]:
nestedList = nestedMatrix()
print(nestedList)

def SummingFunction(listInput):
    totalSum = 0
    for list in range(len(listInput)):
        for numbers in range(len(listInput[list])):
            if listInput[list][numbers] != "?":
                totalSum += listInput[list][numbers]
    return totalSum

SummingFunction(nestedList)

[[1, 2, '?', 4, 5], ['?', 7, 8, '?', 10], [11, '?', 13, 14, '?'], [16, 17, '?', 19, 20], ['?', 22, 23, '?', 25]]


217